store/sqlstore/channel_store.go

```diff
@@ -3132,3 +3401,287 @@ func (s SqlChannelStore) GroupSyncedChannelCount() (int64, *model.AppError) {
+		}
+
+		category.DisplayName = oldCategory.DisplayName
+		if _, err = transaction.UpdateColumns(func(col *gorp.ColumnMap) bool {
+			return col.ColumnName == "DisplayName"
+		}, category); err != nil {
+			return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+
+		// clean previous favorites
+		sql, args, _ := s.getQueryBuilder().Delete("SidebarChannels").Where(sq.Eq{"ChannelId": oldCategory.Channels}).ToSql()
+
+		if _, err = transaction.Exec(sql, args...); err != nil {
+			return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+		}
+
+		if category.Type == model.SidebarCategoryFavorites {
+			// clean previous preference favorites (to support bi-directional sync)
+			sql, args, _ = s.getQueryBuilder().Delete("Preferences").Where(sq.And{
+				sq.Eq{"ChannelId": oldCategory.Channels},
+				sq.Eq{"Preferences.Category": model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL},
+			}).ToSql()
+
+			if _, err = transaction.Exec(sql, args...); err != nil {
+				return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+			}
+		}
+		var channels []interface{}
+		var preferences []interface{}
+		runningOrder := 0
+		for _, channelID := range oldCategory.Channels {
+			channels = append(channels, &model.SidebarChannel{
+				ChannelId:  channelID,
+				CategoryId: category.Id,
+				SortOrder:  int64(runningOrder),
+				UserId:     userId,
+			})
+			preferences = append(preferences, &model.Preference{
+				Name:     channelID,
+				UserId:   userId,
+				Category: model.PREFERENCE_CATEGORY_FAVORITE_CHANNEL,
+				Value:    "true",
+			})
+			runningOrder += model.MinimalSidebarSortDistance
+		}
+
+		if err = transaction.Insert(channels...); err != nil {
+			return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+
+		}
+
+		if category.Type == model.SidebarCategoryFavorites {
+			if err = transaction.Insert(preferences...); err != nil {
+				return nil, model.NewAppError("SqlPostStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.app_error", nil, err.Error(), http.StatusInternalServerError)
+			}
+		}
+	}
+
+	if err = transaction.Commit(); err != nil {
+		return nil, model.NewAppError("SqlChannelStore.UpdateSidebarCategory", "store.sql_channel.sidebar_categories.commit_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
+	}
+
+	return categories, nil
+}
+
+// UpdateSidebarChannelByPreference is called when the Preference table is being updated to keep SidebarCategories in sync
+func (s SqlChannelStore) UpdateSidebarChannelByPreference(preference *model.Preference) *model.AppError {
```

Comment:
Since all of the linking between preferences and SidebarChannels is being done in the store layer, are we going to be able to emit the proper websocket events when one causes the other to change?

Reply:
you are right. i'll reroute it through App layer and add the events